### ISSUE TITLE
BaseLogFile should provide a method to reopen logs

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4255]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#4255|
|Type|enhancement|
|Created|2010-02-04 09:50:43Z|
|Branch|https://github.com/twisted/twisted/tree/reopen-logfile-4255|

When using an external tool to rotate logs like logrotate, we need to ask the twisted log files to be reopened. `BaseLogFile` is missing a method to do that, because it involves accessing the private method `_openFile`.

<details><summary>Searchable metadata</summary>

```
trac-id__4255 4255
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_reopen_logfile_4255 branches/reopen-logfile-4255
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1265277043000000 1265277043000000
changetime__1266793074000000 1266793074000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_2"></a>|@radix commented|
|-|-|

It'd be nice to actually implement a signal handler to reopen log files in twistd itself, as well. Maybe that should be a separate ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

That would be nice, and it probably should be a separate ticket.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [28260]) Branching to 'reopen-logfile-4255'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

This is ready to review! Nothing surprising, I hope.

**Comment by twisted-trac** [comment]
|<img alt="free.ekanayaka's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|free.ekanayaka set owner to @therve|
|-|-|

Nice fix. If it's not already opened it would be nice to file a ticket for the signal handler.

[1]

+        f = open(self.path, "r")
+        self.assertEquals(f.read(), "hello2")
+        f.close()

I'm not sure if trial has it already, but maybe it would be nice to have an "assertFileContent" convenience to perform this kind of assertion.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [28311]) Merge reopen-logfile-4255

Author: therve
Reviewer: free.ekanayaka
Fixes: [#4255](https://github.com/twisted/twisted/issues/4255)

Add a reopen method to BaseLogfile to allow using external log rotation.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `reopened`|
|-|-|

(In [28315]) Revert r28311: regressions under Windows

Reopens [#4255](https://github.com/twisted/twisted/issues/4255)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner<br>@therve set status to `new`|
|-|-|

There were some problems with Windows, but the test should be skipped now.

**Comment by twisted-trac** [comment]
|<img alt="jkakar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|jkakar set owner to @therve|
|-|-|

[1]

+        L[logfile.LogFile.reopen} allows to rename the currently used file and

The L[ needs to be L{ here.


I can't test that the test is skipped on win32, but it looks fine.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [28334]) Merge reopen-logfile-4255

Author: therve
Reviewers: free.ekanayaka, jkakar
Fixes: [#4255](https://github.com/twisted/twisted/issues/4255)

Add a reopen method to BaseLogfile to allow using external log rotation. This
re-merge fix a test failure under Windows.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set status to `reopened`|
|-|-|

(In [28339]) Revert r28334: forget one review comment.

Reopens [#4255](https://github.com/twisted/twisted/issues/4255)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set status to `closed`|
|-|-|

(In [28341]) Merge reopen-logfile-4255

Author: therve
Reviewers: free.ekanayaka, jkakar
Fixes: [#4255](https://github.com/twisted/twisted/issues/4255)

Add a reopen method to BaseLogfile to allow using external log rotation. This
re-merge fix a test failure under Window and a missed review.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
